python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py

```diff
@@ -216,60 +204,97 @@ def plugin_register_ingest(handle, callback, ingest_ref):
-
-
-async def save_data(data):
-    async_ingest.ingest_callback(c_callback, c_ingest_ref, data)
-
-
-# MQTT Message Received Callback
-# The callback for when a PUBLISH message is received from the server.
-def on_message(client, userdata, msg):
-    global _callback_event_loop
-    global _assetName
-    try:
-        if _callback_event_loop is None:
-            asyncio.set_event_loop(asyncio.new_event_loop())
-            _callback_event_loop = asyncio.get_event_loop()
-        # Protobuf message structure
-        sparkplug_payload = sparkplug_b_pb2.Payload()
-        sparkplug_payload.ParseFromString(msg.payload)
-
-        for metric in sparkplug_payload.metrics:
-            value = "Unknown"
-            if metric.HasField("bool_value"):
-                """ bool value cast to int as internal. See FOGL-8067 """
-                value = metric.bool_value
-            elif metric.HasField("float_value"):
-                value = metric.float_value
-            elif metric.HasField("int_value"):
-                value = metric.int_value
-            elif metric.HasField("string_value"):
-                value = metric.string_value
-            # TODO: FOGL- 9198 - Handle other data types
-            if value == "Unknown":
-                _LOGGER.warning("Ignoring metric '{}' due to unknown type. "
-                                "Only supported types are: float, integer, string, bool.".format(metric.name))
-                continue
-            data = {
-                'asset': _assetName,
-                'timestamp': datetime.fromtimestamp(metric.timestamp, tz=timezone.utc
-                                                    ).strftime('%Y-%m-%d %H:%M:%S.%s'),
-                'readings': {metric.name: value}
-            }
-            _callback_event_loop.run_until_complete(save_data(data))
-    except Exception as ex:
-        msg = ("Message payload must comply with spBv1.0 standards. Please ensure that the format and structure of the "
-               "payload adhere to the specified requirements.")
-        _LOGGER.error(ex, msg)
+class MqttSubscriberClient(object):
+    """ mqtt subscriber """
+
+    __slots__ = ['mqtt_client', 'broker_host', 'broker_port', 'username', 'password', 'topic',
+                 'asset_name', 'loop']
+
+    def __init__(self, config):
+        self.mqtt_client = mqtt.Client()
+        self.broker_host = config['url']['value']
+        self.broker_port = int(config['port']['value'])
+        self.username = config['user']['value']
+        self.password = config['password']['value']
+        self.topic = config['topic']['value']
+        self.asset_name = config['assetName']['value']
+
+    def on_connect(self, client, userdata, flags, rc):
+        """ The callback for when the client receives a CONNACK response from the server """
+
+        if self.topic.startswith('spBv1.0'):
+            client.connected_flag = True
```

Comment:
In the current design a single service instance can subscribe to only one message type from a device as following topic namespace is a single config item. 
namespace/group_id/message_type/edge_node_id/[device_id]

Hence we may need multiple service instance

Reply:
Yes, that’s correct! Currently, the plugin does not support subscribing to multiple topics simultaneously. Therefore, we may need to create multiple service instances to handle subscriptions for each topic individually.